### PR TITLE
namespaces: warn if deleting namespaces that have allocs, volumes, vars or quotas associated with them

### LINKED ISSUE #17088: deleting namespaces that are in use by allocs or volumes should error
When a namespace is deleted we check that the namespace has no jobs in a terminal state (ref [`namespace_endpoint.go#L194-L221`](https://github.com/hashicorp/nomad/blob/v1.5.4/nomad/namespace_endpoint.go#L194-L221)) but this isn't really enough to ensure data integrity. In practice, users need to hit the [List Allocations API](https://developer.hashicorp.com/nomad/api-docs/allocations#list-allocations) and [List Volumes API](https://developer.hashicorp.com/nomad/api-docs/volumes#list-volumes) and [List Variables API](https://developer.hashicorp.com/nomad/api-docs/variables#list-variables) to ensure that they don't have any lingering objects that reference the namespace. This is particularly painful for ENT users who have quotas that need to get decremented when we GC the allocation.

Namespaces are an important-enough "container" object and ACL scope that it makes sense to ensure that we never delete a namespace with references to it. Although this operation is probably slow, deleting namespaces is going to naturally be run much less frequently than job or allocation updates will happen.

### PR DESCRIPTION
**Pull request by pkazmierczak**
Fixes #17088 
Relates to #19979
Relates to #19483 
Relates to #19414 

Internal ref: https://hashicorp.atlassian.net/browse/NET-9341

### CODE REVIEW

File: nomad/namespace_endpoint.go [hunk 1]
```diff
@@ -113,15 +123,30 @@ func (n *Namespace) DeleteNamespaces(args *structs.NamespaceDeleteRequest, reply
...
-	// this region and all federated regions
 	var mErr multierror.Error
 	for _, ns := range args.Namespaces {
-		nonTerminal, err := n.nonTerminalNamespaces(args.AuthToken, ns)
-		if err != nil {
-			_ = multierror.Append(&mErr, err)
-		} else if len(nonTerminal) != 0 {
-			_ = multierror.Append(&mErr, fmt.Errorf("namespace %q has non-terminal jobs in regions: %v", ns, nonTerminal))
+		// do a check across jobs, allocations, volumes and variables to make sure we're
+		// not leaving any objects associated with the namespace hanging
```
[the remark this answers is not in the document]
Comment (pkazmierczak) [reply nomad/namespace_endpoint.go:129]: oh I didn't notice we weren't checking 🤦

File: nomad/namespace_endpoint.go [hunk 2]
```diff
@@ -113,15 +123,30 @@ func (n *Namespace) DeleteNamespaces(args *structs.NamespaceDeleteRequest, reply
...
+			{n.namespaceNoAssociatedQuotasLocally, nil, "namespace %q has quotas associated with it: %v"},
+		}
+
+		for _, object := range objects {
+			leftovers, err := n.nonTerminalObjectsInNS(args.AuthToken, ns, object.localCheckFunc, object.remoteCheckFunc)
+			if err != nil {
+				_ = multierror.Append(&mErr, err)
+			} else if len(leftovers) != 0 {
+				_ = multierror.Append(&mErr, fmt.Errorf(object.errorMsg, ns, leftovers))
+			}
```
[the remark this answers is not in the document]
Comment (pkazmierczak) [reply nomad/namespace_endpoint.go]: addressed in [88dc567](https://github.com/hashicorp/nomad/pull/23499/commits/88dc567d554ba5c794625ef39a5f82419179b71b)

File: nomad/namespace_endpoint.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pkazmierczak) [reply nomad/namespace_endpoint.go:132]: addressed in [88dc567](https://github.com/hashicorp/nomad/pull/23499/commits/88dc567d554ba5c794625ef39a5f82419179b71b)

File: nomad/namespace_endpoint.go [hunk 3]
```diff
@@ -140,16 +165,25 @@ func (n *Namespace) DeleteNamespaces(args *structs.NamespaceDeleteRequest, reply
...
 	regions := n.srv.Regions()
 	thisRegion := n.srv.Region()
 	terminal := make([]string, 0, len(regions))
 
 	// Check if this region is terminal
-	localTerminal, err := n.namespaceTerminalLocally(namespace)
+	snap, err := n.srv.fsm.State().Snapshot()
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (pkazmierczak) [reply nomad/namespace_endpoint.go]: addressed in [88dc567](https://github.com/hashicorp/nomad/pull/23499/commits/88dc567d554ba5c794625ef39a5f82419179b71b)

File: nomad/namespace_endpoint.go [hunk 4]
```diff
@@ -214,20 +325,103 @@ func (n *Namespace) namespaceTerminalInRegion(authToken, namespace, region strin
...
+		return false, fmt.Errorf("unexpectedly did not forward Alloc.List to region %q", region)
+	} else if err != nil {
+		return false, err
+	}
+
+	for _, alloc := range allocResp.Allocations {
+		if !slices.Contains(terminalAllocationStatuses, alloc.ClientStatus) {
+			return false, nil
+		}
+	}
```
[the remark this answers is not in the document]
Comment (pkazmierczak) [reply nomad/namespace_endpoint.go]: good catch, I didn't notice that method. [88dc567](https://github.com/hashicorp/nomad/pull/23499/commits/88dc567d554ba5c794625ef39a5f82419179b71b) fixes this.

**Review by tgross — CHANGES_REQUESTED** [review]

File: nomad/namespace_endpoint.go [hunk 5]
```diff
@@ -113,15 +113,40 @@ func (n *Namespace) DeleteNamespaces(args *structs.NamespaceDeleteRequest, reply
...
+	snap, err := n.srv.fsm.State().Snapshot()
+	if err != nil {
+		return err
+	}
+
 	var mErr multierror.Error
 	for _, ns := range args.Namespaces {
-		nonTerminal, err := n.nonTerminalNamespaces(args.AuthToken, ns)
+		// make sure this namespace exists before we start making costly checks
+		_, err := snap.NamespaceByName(nil, ns)
```
[the remark this answers is not in the document]
Comment (pkazmierczak) [reply nomad/namespace_endpoint.go:126]: ugh, I saw https://github.com/hashicorp/nomad/blob/88dc567d554ba5c794625ef39a5f82419179b71b/nomad/state/state_store.go#L6802 and didn't look at `txn.FirstWatch` 🤦

[the remark this answers is not in the document]
Comment (pkazmierczak) [reply nomad/namespace_endpoint.go]: fixed in [b5af6ce](https://github.com/hashicorp/nomad/pull/23499/commits/b5af6ce7a71e3ca1e4e898dedd5e1e687be83dba)

File: nomad/namespace_endpoint.go [hunk 6]
```diff
@@ -162,50 +193,131 @@ func (n *Namespace) nonTerminalNamespaces(authToken, namespace string) ([]string
...
-// jobs in the given region .
-func (n *Namespace) namespaceTerminalInRegion(authToken, namespace, region string) (bool, error) {
-	req := &structs.JobListRequest{
+// namespaceNoAssociatedQuotasLocally returns true if there are no quotas
+// associated with this namespace in the local region
+func (n *Namespace) namespaceNoAssociatedQuotasLocally(namespace string, snap *state.StateSnapshot) (bool, error) {
+	ns, err := snap.NamespaceByName(nil, namespace)
+	if err != nil {
+		return false, err
+	}
```
[the remark this answers is not in the document]
Comment (pkazmierczak) [reply nomad/namespace_endpoint.go]: fixed in [b5af6ce](https://github.com/hashicorp/nomad/pull/23499/commits/b5af6ce7a71e3ca1e4e898dedd5e1e687be83dba)

**Review by tgross — APPROVED** [review]
LGTM!

**Pull request merged**